enhancements/dns/nodeservice-etc-hosts-injection.md

```diff
@@ -0,0 +1,422 @@
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: DNS
+metadata:
+  name: default
+spec:
+  nodeServices:
+  - namespace: my-custom-namespace
+    name: my-custom-registry
+```
+
+#### Story 2: Cluster provided infrastructure services
+
+As a platform team managing critical infrastructure services that need to be available to node scoped processes we want /etc/hosts entries to be populated to provide these entries 
+
+Example:
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: DNS
+metadata:
+  name: default
+spec:
+  nodeServices:
+  - namespace: coredump
+    name: datastore
+```
+
+### Implementation Details/Notes/Constraints
+
+#### DNS Node Resolver Implementation
+
+The DNS node resolver (running as a DaemonSet on all nodes) will:
+
+1. The cluster-dns-operator watches the `DNS` custom resource for changes to the `nodeServices` field
+2. For each service specified in `nodeServices`, the operator constructs a list of service names, the internal
+   registry service `image-registry.openshift-image-registry.svc` is always added to the list
+   - Format: Service names in `service-name.namespace.svc` format
+   - Example: `datastore.coredump.svc`
+3. The operator populates the node-resolver DaemonSet `SERVICES` environment variable with the list of service names
+   - Format: Comma-separated and/or space-separated list of service names in `name.namespace.svc` format
+   - Example: `SERVICES="image-registry.openshift-image-registry.svc, datastore.coredump.svc"`
+   - The script appends `.${CLUSTER_DOMAIN}` when performing DNS lookups to create the full FQDN
+4. The DNS node resolver pods read the `SERVICES` environment variable on startup
+5. The `update-node-resolver.sh` script runs on each node and:
+   - Parses the list of service names from the `SERVICES` environment variable (comma/space-separated)
+   - For each service, performs DNS lookups using `dig` to resolve the ClusterIP
+     - Tries multiple resolution methods: IPv4, IPv6, and TCP fallback (for environments like Kuryr that don't support UDP)
+     - Uses the cluster's nameserver (`${NAMESERVER}`) and cluster domain (`${CLUSTER_DOMAIN}`)
+   - Updates `/etc/hosts` with entries in the format: `<ClusterIP> <service-name>.<namespace> <service-name>.<namespace>.<cluster-domain> # openshift-generated-node-resolver`
+   - Runs in a 60-second loop to continuously detect and apply IP changes
+   - Only updates `/etc/hosts` if valid IPs are resolved (preserves stale entries if DNS/API is unavailable)
+   - Uses atomic file operations to prevent corruption
+6. When the `nodeServices` list changes in the DNS CR, the operator updates the DaemonSet `SERVICES` environment variable, triggering a rolling restart to pick up the new service list
+
+Steps 1-4 are the only steps modified by this enhancement.
+
+#### Validation
+
+The following validation rules will be enforced:
+
+- `namespace` and `name` fields must not be empty
+- Service names must conform to DNS label standards (RFC 1123)
```

Comment:
It's actually DNS1035 label, https://github.com/kubernetes/kubernetes/blob/c7b6dfb144aa24adaf443f9a3682f971e4f21140/pkg/apis/core/validation/validation.go#L271